proposals/0252-callable.md

```diff
@@ -0,0 +1,611 @@
+# Introduce callables
+
+* Proposal: SE-NNNN
+* Authors: [Richard Wei](https://github.com/rxwei), [Dan Zheng](https://github.com/dan-zheng)
+* Review Manager: TBD
+* Status: Implementation in progress
+* Implementation: https://github.com/apple/swift/pull/23517
+
+## Introduction
+
+Note: throughout this document, let "call-syntax" refer to the syntax of applying a function to arguments: `f(x, y, z)`.
+
+This proposal introduces [callables](https://en.wikipedia.org/wiki/Callable_object) to Swift. Callables are values that define function-like behavior and can be applied using function application syntax.
+
+In a nutshell, we propose to introduce a new declaration syntax with the keyword `call` :
+
+```swift
+struct Adder {
+    var base: Int
+    call(_ x: Int) -> Int {
+        return base + x
+    }
+}
+```
+
+Values that have a `call` member can be applied like functions, forwarding arguments to the `call` member.
+
+```swift
+let add3 = Adder(base: 3)
+add3(10) // => 13
+```
+
+## Motivation
+
+Currently, in Swift, only a few kinds of values are syntactically callable:
+- Values with function types.
+- Type names (e.g. `T` can be called like `T(...)`, which is desugared to `T.init(...)`).
+- Values with a `@dynamicCallable` type.
+
+However, callable syntax can also be useful for other values, primarily those that behave like functions. This includes:

```

Comment:
Unify terminology: rename "callable syntax"and "call-syntax" to "call syntax"?
There are four mentions of "callable syntax".

Reply:
Done.